### ISSUE TITLE
Reusing pieces of models

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#844]
Most of DeepChem is based on the assumption that you want to create a model, then train that model, then make predictions based on that trained model.  But sometimes that isn't what you want.  Rather than making predictions with the trained model directly, you might want to reuse part or all of it as one piece of a larger model.

Here's an example I'm dealing with right now.  I want to start with a SeqToSeq model and train it as an autoencoder.  Then I want to take just the encoder portion of that trained model, and use it as part of a policy for RL.  I can hack something together that will do that, but it will be ugly, and it seems to me this is a common problem.  Do we want to consider ways to make this sort of thing easier?

**Comment by peastman** [comment]
Since no one has commented on this, let me offer a candidate proposal.

1. We'll add an optional field to Layer that stores the initial values of all its variables.  When you create a TensorGraph containing that layer, it starts by setting the variables to those values.  For layers that already have configurable initializers, this will override them.  The idea is that this lets a Layer object describe a *pretrained* layer.

2. We might consider adding an option to `TensorGraph.save()` that tells it to include the values of all trainable variables in the saved layers.  That would allow you to store a complete trained model, both the model structure and the parameter values, in a single file.

3. We'll create a function to clone a tree of Layer objects.  You call it on a single Layer, and it recursively creates a copy of that Layer and every parent Layer it depends on.  You can optionally specify a TensorGraph, and the cloned Layers will include the values of all trainable variables from that graph.  This lets you clone a piece of a trained model and have the variables get copied with it.

4. We also might consider adding a StopGradient layer that passes its input through `tf.stop_gradient()`.  That way you could incorporate a pretrained model into a larger model, and tell it that the pretrained layers shouldn't be optimized any further.  It also could be useful in other contexts, like GANs.

**Comment by rbharath** [comment]
As a counterproposal, I suggest we implement Keras's functional API. In this API, models themselves can be invoked as layers:

https://keras.io/getting-started/functional-api-guide/

This should allow for an elegant solution to the resuse problem and has some buy-in already from larger deep learning community as a standard API

**Comment by peastman** [comment]
I don't see how that solves the problem.  In the example above I don't want to reuse the whole model, just one piece of it.  More generally, pretraining involves training a set of layers in the context of one model, then reusing those trained layers in the context of a different model.  It requires fine control to be able to pull out just the specific pieces you want.

Also, I think you're only suggesting that as an alternative to number 3 above?  I don't think it has much to do with numbers 1, 2, and 4.

**Comment by rbharath** [comment]
For Seq2Seq, in the functional API, we would define it as the composition of two models, an encoder model and a decoder model. Then to reuse the encoder, you can just pull out the encoder subcomponent of the Seq2Seq model.

For the pretraining problem in general, you'd just define submodels in places where you think you might want to pull out pieces later. In this view, models are basically generalized layers, so it allows for elegant decompositions.

I think the functional API is a healthy high level API evolution in general. If we decide it's worthwhile, we should think about spec-ing out the concrete changes needed

**Comment by peastman** [comment]
I think we're talking about orthogonal things.  The "submodels" you suggest might be a useful abstraction to introduce, but they're a different sort of entity from "models" in DeepChem.  A TensorGraph model specifies a loss function and an optimizer.  It's associated with a particular Tensorflow graph, and it owns a session that it uses to store its variables.  None of those would be true of submodels.

I think what you're talking about is just having a "metalayer" that groups a bunch of layers together and lets you work with it as a single layer?  That's potentially useful, but it doesn't replace the need for any of the things I suggested above.  You still need a way to copy layers from one model to another while preserving their variables, whether or not those layers have been grouped into a submodel.

What I'm talking about is something much closer to Tensorflow's [graph_replace()](https://www.tensorflow.org/api_docs/python/tf/contrib/graph_editor/graph_replace) function.  A way of telling it, "Make a copy of these layers while hooking them up to new inputs."

**Comment by rbharath** [comment]
Ah, I see your point now. I agree we're talking about orthogonal issues. Yes, I think the "metalayer" is the correct replacement for 3 above (since we might not want to clone the entire tree but rather parts of it), but the overall proposal seems sound to me.

**Comment by proteneer** [comment]
I'm running into this problem as well. What was the end conclusion of this?

**Comment by peastman** [comment]
Items 1, 3, and 4 above were implemented in #850.

**Comment by rbharath** [comment]
Closing this issue since #850 implemented
